### PR TITLE
Better shop ordering

### PR DESCRIPTION
**Pull request by warp-core**
**Feature:** This PR implements the feature request detailed and discussed in issue #7046

## Feature Details
Outfits and ships can be given a 'series' attribute and 'index' number. In the shop, they will be grouped (within categories) by series, then sorted by index, then by name alphabetically (for outfits with the same category, series, and index).

Series will be sorted based on precedence values assigned separetly, in a similar way to the ordering of the already existing categories for outfits and ships. Series with the same precedence will be ordered alphabetically.

~~Should I not update any of the data in this PR and leave that for another one? Or, update it all?~~

[Better shop ordering test.txt](https://github.com/endless-sky/endless-sky/files/10255078/Better.shop.ordering.test.txt)

Put this in a plugin and it should work as though it were in the base game data, maybe.

## UI Screenshots
Ok, it works now, I think. I've only added the new ordering tags to human engines here.
image

## Usage Examples
```
category "series"
	# Misc
	"Afterburners: Human" 1000
	"Afterburners: Remnant" 1010
	"Combo engines: Human" 10
	"Combo engines: Hai" 20
	"Combo engines: Remnant" 30
```
```
outfit "Greyhound Plasma Steering"
	category "Engines"
	series "Steering: Human: Plasma"
	index 20
```

## Testing Done
Start the game, go to the outfitter (probably use an all content plugin).

## Performance Impact
All outfits/ships are sorted upon opening the respective shop panel.

### CODE REVIEW

File: source/CategoryList.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
+/* CategoryList.cpp
+Copyright (c) 2022 by Amazinite
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/CategoryList.cpp:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```
The standard copyright header was changed recently.

File: source/CategoryList.h [hunk 2]
```diff
@@ -0,0 +1,78 @@
+/* CategoryList.h
+Copyright (c) 2022 by Amazinite
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/CategoryList.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

File: source/comparators/BySeriesAndIndex.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
+/* comparators/BySeriesAndIndex.h
+Copyright (c) 2022 by warp-core
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
```
Comment (tibetiroka) [inline source/comparators/BySeriesAndIndex.h:10]:
```suggestion
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program. If not, see <https://www.gnu.org/licenses/>.
```

File: data/human/engines.txt [hunk 4]
```diff
@@ -90,6 +90,8 @@ effect "ionic afterburner"
 outfit "X1050 Ion Engines"
 	plural "X1050 Ion Engines"
 	category "Engines"
+	series "Combo engines: Human"
+	index 10
```
Comment (warp-core) [inline data/human/engines.txt:94]:
```suggestion
```

File: source/comparators/BySeriesAndIndex.h [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+#include "../CategoryList.h"
+#include "../CategoryTypes.h"
+#include "../GameData.h"
+#include "../Outfit.h"
+#include "../Ship.h"
+
+namespace {
+	bool Helper(const Outfit &a, const Outfit &b, const std::string &nameA, const std::string &nameB)
+	{
+		CategoryList series = GameData::GetCategory(CategoryType::SERIES);
```
Comment (warp-core) [inline source/comparators/BySeriesAndIndex.h:28]:
```suggestion
		static const CategoryList &series = GameData::GetCategory(CategoryType::SERIES);
```

**Comment by warp-core** [comment]
Nah, it won't. Unless it does???

File: source/GameData.h [hunk 6]
```diff
@@ -160,7 +161,7 @@ class GameData {
 	// Strings for combat rating levels, etc.
 	static const std::string &Rating(const std::string &type, int level);
 	// Strings for ship, bay type, and outfit categories.
```
Comment (Amazinite) [inline source/GameData.h:163]:
```suggestion
	// Collections for ship, bay type, outfit, and other categories.
```

File: source/GameData.cpp [hunk 7]
```diff
@@ -778,7 +779,7 @@ const string &GameData::Rating(const string &type, int level)
 
 
 // Strings for ship, bay type, and outfit categories.
```
Comment (Amazinite) [inline source/GameData.cpp:781]:
```suggestion
// Collections for ship, bay type, outfit, and other categories.
```

File: source/Outfit.h [hunk 8]
```diff
@@ -105,6 +107,8 @@ class Outfit : public Weapon {
 	std::string displayName;
 	std::string pluralName;
 	std::string category;
+	std::string series;
+	int index;
```
Comment (Amazinite) [inline source/Outfit.h:111]: It's not obvious from their names what these are used for, so I think a comment is appropriate.
```suggestion
	// The series that this outfit is a part of and its index within that series.
	// Used for sorting within shops.
	std::string series;
	int index;
```

File: source/CategoryList.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+			byName.insert(pair<const string, Category>(cat.name, cat));
+		}
+
+	}
+}
+
+
+
+// Sort the CategoryList. Categories are sorted by precedence. If multiple categories
+//  share the same precedence then they are sorted alphabetically.
```
Comment (Amazinite) [inline source/CategoryList.cpp:63]:
```suggestion
// share the same precedence then they are sorted alphabetically.
```

File: source/CategoryList.cpp [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+		[&name](const Category &c) noexcept -> bool { return name == c.name; });
+	return it != list.end();
+}
+
+
+
+const CategoryList::Category CategoryList::GetCategory(const string &name) const
+{
+	if(byName.count(name))
+		return byName.at(name);
```
Comment (Amazinite) [inline source/CategoryList.cpp:84]: We normally use `map::find` for this, as I believe that doing `count` then `at` is essentially searching the map twice.

Comment (warp-core) [inline source/CategoryList.cpp:84]:
```suggestion
	auto it = byName.find(name);
	if(it != byName.end())
		return it->second;
```

File: source/CategoryList.h [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+		Category(std::string name, int precedence) : name(name), precedence(precedence) {}
+		const std::string &Name() const { return name; }
+		const bool operator<(Category other) const { return SortHelper(*this, other); }
+		const bool operator()(Category &a, Category &b) const { return SortHelper(a, b); }
+
+	private:
+		friend class CategoryList;
+		std::string name;
+		int precedence = 0;
+		const bool SortHelper(const Category &a, const Category &b) const;
```
Comment (warp-core) [inline source/CategoryList.h:50]: Methods declared separately to fields.
This is also independent of the instances.
```suggestion
	private:
		static const bool SortHelper(const Category &a, const Category &b) const;


	private:
		friend class CategoryList;
		std::string name;
		int precedence = 0;
```

**Pull request merged**